values.go

```diff
@@ -258,104 +181,298 @@ func NormalizeType(value interface{}) Value {
+}
 
-	if (left == nil && right != nil) ||
-		(left != nil && right == nil) {
-		return false
-	}
+type Comparison int
 
-	switch left := left.(type) {
-	case Null:
-		_, ok := right.(Null)
-		if !ok {
-			return false
-		}
-		return true
+const (
+	LessThan    Comparison = -1
+	Equal       Comparison = 0
+	GreaterThan            = 1
+)
 
-	case Phantom:
-		_, ok := right.(Phantom)
-		if !ok {
-			return false
+func Compare(x, y Value) (Comparison, error) {
+	switch x.GetType() {
+	case TypeInt:
+		if y.GetType() != TypeInt {
+			return 0, errors.Errorf("type mismatch between values")
 		}
-		return true
 
-	case Int:
-		right, ok := right.(Int)
-		if !ok {
-			return false
-		}
-		return left == right
+		x := x.AsInt()
+		y := y.AsInt()
 
-	case Float:
-		right, ok := right.(Float)
-		if !ok {
-			return false
+		if x == y {
+			return 0, nil
+		} else if x < y {
+			return -1, nil
 		}
-		return left == right
 
-	case Bool:
-		right, ok := right.(Bool)
-		if !ok {
-			return false
+		return 1, nil
+	case TypeFloat:
+		if y.GetType() != TypeFloat {
+			return 0, errors.Errorf("type mismatch between values")
 		}
-		return left == right
+		x := x.AsFloat()
+		y := y.AsFloat()
 
-	case String:
-		right, ok := right.(String)
-		if !ok {
-			return false
+		if x == y {
+			return 0, nil
+		} else if x < y {
+			return -1, nil
 		}
-		return left == right
 
-	case Time:
-		right, ok := right.(Time)
-		if !ok {
-			return false
+		return 1, nil
+	case TypeString:
+		if y.GetType() != TypeString {
+			return 0, errors.Errorf("type mismatch between values")
 		}
 
-		return left.AsTime().Equal(right.AsTime())
+		x := x.AsString()
+		y := y.AsString()
 
-	case Duration:
-		right, ok := right.(Duration)
-		if !ok {
-			return false
+		if x == y {
+			return 0, nil
+		} else if x < y {
+			return -1, nil
 		}
-		return left == right
 
-	case Tuple:
-		right, ok := right.(Tuple)
-		if !ok {
-			return false
+		return 1, nil
+	case TypeTime:
+		if y.GetType() != TypeTime {
+			return 0, errors.Errorf("type mismatch between values")
 		}
-		if len(left) != len(right) {
-			return false
+
+		x := x.AsTime()
+		y := y.AsTime()
+
+		if x == y {
+			return 0, nil
+		} else if x.Before(y) {
+			return -1, nil
 		}
-		for i := range left {
-			if !AreEqual(left[i], right[i]) {
-				return false
-			}
+
+		return 1, nil
+	case TypeBool:
+		if y.GetType() != TypeBool {
+			return 0, errors.Errorf("type mismatch between values")
 		}
-		return true
 
-	case Object:
-		right, ok := right.(Object)
-		if !ok {
-			return false
+		x := x.AsBool()
+		y := y.AsBool()
+
+		if x == y {
+			return 0, nil
+		} else if !x && y {
+			return -1, nil
 		}
-		if len(left) != len(right) {
-			return false
+
+		return 1, nil
+
+	case TypeNull, TypePhantom, TypeDuration, TypeTuple, TypeObject:
+		return 0, errors.Errorf("unsupported type in sorting")
+	}
+
+	panic("unreachable")
+}
+
+func ZeroValue() Value {
+	return Value{}
+}
+
+func (v Value) AsInt() int {
+	return int(v.GetInt())
```

Comment:
why there (and only there) is type casting?

Reply:
Because getint is int64